src/Http/WebUtilities/src/FormPipeReader.cs

```diff
@@ -0,0 +1,346 @@
+            ref KeyValueAccumulator accumulator,
+            bool isFinalBlock)
+        {
+            var sequenceReader = new SequenceReader<byte>(buffer);
+            var consumed = sequenceReader.Position;
+            var equalsDelimiter = GetEqualsForEncoding();
+            var andDelimiter = GetAndForEncoding();
+
+            while (!sequenceReader.End)
+            {
+                // TODO seems there is a bug with TryReadTo (advancePastDelimiter: true). It isn't advancing past the delimiter on second read.
+                if (!sequenceReader.TryReadTo(out ReadOnlySequence<byte> key, equalsDelimiter, advancePastDelimiter: false) ||
+                    !sequenceReader.IsNext(equalsDelimiter, true))
+                {
+                    break;
+                }
+
+                if (!sequenceReader.TryReadTo(out ReadOnlySequence<byte> value, andDelimiter, false) ||
+                    !sequenceReader.IsNext(andDelimiter, true))
+                {
+                    if (!isFinalBlock)
+                    {
+                        break;
+                    }
+
+                    value = buffer.Slice(sequenceReader.Position);
+
+                    sequenceReader.Advance(value.Length);
+                }
+
+                // Need to call ToArray if the key/value spans multiple segments 
+                var decodedKey = GetDecodedStringFromReadOnlySequence(key);
+                var decodedValue = GetDecodedStringFromReadOnlySequence(value);
+
+                AppendAndVerify(ref accumulator, decodedKey, decodedValue);
+
+                consumed = sequenceReader.Position;
+            }
+
+            buffer = buffer.Slice(consumed);
+        }
+
+        private string GetDecodedStringFromReadOnlySequence(ReadOnlySequence<byte> ros)
+        {
+            if (ros.IsSingleSegment)
+            {
+                return GetDecodedString(ros.First.Span);
+            }
+
+            if (ros.Length < StackAllocThreshold)
+            {
+                Span<byte> buffer = stackalloc byte[(int)ros.Length];
+                ros.CopyTo(buffer);
+                return GetDecodedString(buffer);
+            }
+            else
+            {
+                var byteArray = ArrayPool<byte>.Shared.Rent((int)ros.Length);
+                Span<byte> buffer = byteArray.AsSpan(0, (int)ros.Length);
+
+                ros.CopyTo(buffer);
+                var decodedString = GetDecodedString(buffer);
+
+                ArrayPool<byte>.Shared.Return(byteArray);
+                return decodedString;
+            }
+        }
+
+        // Check that key/value constraints are met and appends value to accumulator.
+        private KeyValueAccumulator AppendAndVerify(ref KeyValueAccumulator accumulator, string decodedKey, string decodedValue)
+        {
+            if (decodedKey.Length > KeyLengthLimit)
+            {
+                throw new InvalidDataException($"Form key or value length limit {KeyLengthLimit} exceeded.");
```

Comment:
Wouldn't that operate on the byte array length rather than the decoded key/value length though? I don't think there is a better place besides this.

Reply:
No, this is a blocker. What happens if I send a 2gb form key or value? You'd buffer the whole thing.

Reply:
Good point. The fields like ValueCountLimit, KeyLengthLimit, etc are ambiguous to begin with but they were originally counting the number of characters instead of bytes. Should we just start counting bytes instead and change the default based on encoding?

Reply:
Hmm. For ascii, UTF-16, and UTF32 the ratio of bytes to chars is fixed. For UTF-8 that doesn't work though, and it's twice as bad with the percent encoding. Counting bytes is always going to be the most restrictive option but I don't think UTF-8 gives you a good alternative.

Changing the default based on the encoding would be problematic for the same reason, you don't know how much to expect for UTF-8.

Reply:
How about we use the max bytes per character as a safety net. We check both in both places (after indexOf and the decoded string).

Reply:
Let just make a breaking change to limit the bytes vs chars. That way we can just check as part of the parsing instead of after

Reply:
Max bytes per char for UTF-8 is a 4x multiplier over the number of characters usually used. Counting the actual bytes is better, it will be ascii most of the time. No need to check the string, it's not going to get longer after decoding.

Reply:
Agreed let’s nust check the bytes

Reply:
Throwing for bytes now. It's a bit tacky because I need to check whether to throw in two places. for each key/value 😢